File: src/lib/api/bankTransaction.ts
/**
 * 银行流水统一台账 API
 */

import { apiDelete, apiGet, apiPost, apiPut } from '@/lib/api/client.ts'

export type BankTransactionType =
  | 'statement_import'
  | 'payment_register'
  | 'collection_register'

export interface BankTransactionRow {
  id: string
  type: BankTransactionType
  trade_date: string | null
  bank_account: string | null
  payer_name: string | null
  payer_account: string | null
  payer_bank_name: string | null
  payee_name: string | null
  payee_account: string | null
  payee_bank_name: string | null
  amount: string | number | null
  income_amount: string | number | null
  expense_amount: string | number | null
  currency: string | null
  transaction_no: string | null
  instruction_no: string | null
  summary: string | null
  purpose: string | null
  remark: string | null
  status: string | null
  raw_text: string | null
  attachment_url: string | null
  created_at: string
  updated_at: string
}

export type BankTransactionCreateBody = Omit<
  BankTransactionRow,
  'id' | 'created_at' | 'updated_at'
>

export type BankTransactionUpdateBody = Partial<
  Omit<BankTransactionRow, 'id' | 'created_at' | 'updated_at'>
>

export interface BankTransactionListResponse {
  items: BankTransactionRow[]
  total: number
}

const PATH = '/api/bank-transactions'

export interface GetBankTransactionsParams {
  q?: string
  type?: BankTransactionType | ''
  date_from?: string
  date_to?: string
  amount_min?: string
  amount_max?: string
  limit?: number
  offset?: number
}

export async function getBankTransactions(
  params: GetBankTransactionsParams = {}
): Promise<BankTransactionListResponse> {
  const sp = new URLSearchParams()
  if (params.q?.trim()) sp.set('q', params.q.trim())
  if (params.type) sp.set('type', params.type)
  if (params.date_from?.trim()) sp.set('date_from', params.date_from.trim())
  if (params.date_to?.trim()) sp.set('date_to', params.date_to.trim())
  if (params.amount_min?.trim()) sp.set('amount_min', params.amount_min.trim())
  if (params.amount_max?.trim()) sp.set('amount_max', params.amount_max.trim())
  if (params.limit != null) sp.set('limit', String(params.limit))
  if (params.offset != null) sp.set('offset', String(params.offset))
  const qs = sp.toString()
  return apiGet<BankTransactionListResponse>(`${PATH}${qs ? `?${qs}` : ''}`)
}

export async function getBankTransactionDetail(id: string): Promise<BankTransactionRow> {
  return apiGet<BankTransactionRow>(`${PATH}/${encodeURIComponent(id)}`)
}

export async function createBankTransaction(
  body: BankTransactionCreateBody
): Promise<BankTransactionRow> {
  return apiPost<BankTransactionRow>(PATH, body)
}

export async function updateBankTransaction(
  id: string,
  body: BankTransactionUpdateBody
): Promise<BankTransactionRow> {
  return apiPut<BankTransactionRow>(`${PATH}/${encodeURIComponent(id)}`, body)
}

export async function deleteBankTransaction(id: string): Promise<void> {
  return apiDelete(`${PATH}/${encodeURIComponent(id)}`)
}

File: backend/app/schemas/bank_transaction.py
"""银行流水台账 API模型。"""

from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, Field

BankTransactionType = Literal["statement_import", "payment_register", "collection_register"]


class BankTransactionCreate(BaseModel):
    type: BankTransactionType
    trade_date: str | None = None
    bank_account: str | None = None
    payer_name: str | None = None
    payer_account: str | None = None
    payer_bank_name: str | None = None
    payee_name: str | None = None
    payee_account: str | None = None
    payee_bank_name: str | None = None
    amount: Decimal | None = None
    income_amount: Decimal | None = None
    expense_amount: Decimal | None = None
    currency: str | None = Field(default="CNY")
    transaction_no: str | None = None
    instruction_no: str | None = None
    summary: str | None = None
    purpose: str | None = None
    remark: str | None = None
    status: str | None = None
    raw_text: str | None = None
    attachment_url: str | None = None


class BankTransactionUpdate(BaseModel):
    type: BankTransactionType | None = None
    trade_date: str | None = None
    bank_account: str | None = None
    payer_name: str | None = None
    payer_account: str | None = None
    payer_bank_name: str | None = None
    payee_name: str | None = None
    payee_account: str | None = None
    payee_bank_name: str | None = None
    amount: Decimal | None = None
    income_amount: Decimal | None = None
    expense_amount: Decimal | None = None
    currency: str | None = None
    transaction_no: str | None = None
    instruction_no: str | None = None
    summary: str | None = None
    purpose: str | None = None
    remark: str | None = None
    status: str | None = None
    raw_text: str | None = None
    attachment_url: str | None = None


class BankTransactionRead(BaseModel):
    model_config = {"from_attributes": True}

    id: str
    type: str
    trade_date: str | None
    bank_account: str | None
    payer_name: str | None
    payer_account: str | None
    payer_bank_name: str | None
    payee_name: str | None
    payee_account: str | None
    payee_bank_name: str | None
    amount: Decimal | None
    income_amount: Decimal | None
    expense_amount: Decimal | None
    currency: str | None
    transaction_no: str | None
    instruction_no: str | None
    summary: str | None
    purpose: str | None
    remark: str | None
    status: str | None
    raw_text: str | None
    attachment_url: str | None
    created_at: datetime
    updated_at: datetime


class BankTransactionListResponse(BaseModel):
    items: list[BankTransactionRead]
    total: int
